File: majorProject/src/main/java/com/aocc/majorproject/GameSession.java
package com.aocc.majorproject;

/**
 * Per-playthrough game state. Replaces static fields previously scattered on {@link GameScreen}.
 */
public class GameSession {

    private static final float INITIAL_ENEMY_COUNTER = 300f;

    private final Player player;
    private float updateCount;
    private float enemyCounter = INITIAL_ENEMY_COUNTER;
    private int score;
    private boolean scoreUploaded;
    private int speed;
    private boolean gameOverFlag;

    public GameSession() {
        player = new Player(this);
    }

    public Player getPlayer() {
        return player;
    }

    public float getUpdateCount() {
        return updateCount;
    }

    public void addUpdateCount(float step) {
        updateCount += step;
    }

    public void subtractUpdateCount(float amount) {
        updateCount -= amount;
    }

    public float getEnemyCounter() {
        return enemyCounter;
    }

    public void addEnemyCounter(float step) {
        enemyCounter += step;
    }

    public void resetEnemyCounter() {
        enemyCounter = 0;
    }

    public int getScore() {
        return score;
    }

    public void addScore(int points) {
        score += points;
    }

    public boolean isScoreUploaded() {
        return scoreUploaded;
    }

    public void setScoreUploaded(boolean scoreUploaded) {
        this.scoreUploaded = scoreUploaded;
    }

    public int getSpeed() {
        return speed;
    }

    public void incrementSpeed() {
        speed++;
    }

    public boolean isGameOverFlag() {
        return gameOverFlag;
    }

    public void setGameOverFlag(boolean gameOverFlag) {
        this.gameOverFlag = gameOverFlag;
    }

    /** Resets counters for a new run while keeping the same screen instance. */
    public void resetForNewRun() {
        updateCount = 0;
        enemyCounter = INITIAL_ENEMY_COUNTER;
        score = 0;
        scoreUploaded = false;
        speed = 0;
        gameOverFlag = false;
    }
}

File: majorProject/src/main/kotlin/com/aocc/majorproject/GameSession.kt
package com.aocc.majorproject

/**
 * Per-playthrough game state. Replaces static fields previously scattered on [GameScreen].
 */
class GameSession {

    private val player = Player(this)
    private var updateCount = 0f
    private var enemyCounter = INITIAL_ENEMY_COUNTER
    private var score = 0
    private var scoreUploaded = false
    private var speed = 0
    private var gameOverFlag = false

    fun getPlayer(): Player = player

    fun getUpdateCount(): Float = updateCount

    fun addUpdateCount(step: Float) {
        updateCount += step
    }

    fun subtractUpdateCount(amount: Float) {
        updateCount -= amount
    }

    fun getEnemyCounter(): Float = enemyCounter

    fun addEnemyCounter(step: Float) {
        enemyCounter += step
    }

    fun resetEnemyCounter() {
        enemyCounter = 0f
    }

    fun getScore(): Int = score

    fun addScore(points: Int) {
        score += points
    }

    fun isScoreUploaded(): Boolean = scoreUploaded

    fun setScoreUploaded(scoreUploaded: Boolean) {
        this.scoreUploaded = scoreUploaded
    }

    fun getSpeed(): Int = speed

    fun incrementSpeed() {
        speed++
    }

    /** Debug-popup override for the speed ramp (0..[com.aocc.framework.GameConstants.SPEED_RAMP_MAX]). */
    fun setSpeed(speed: Int) {
        this.speed = speed
    }

    fun isGameOverFlag(): Boolean = gameOverFlag

    fun setGameOverFlag(gameOverFlag: Boolean) {
        this.gameOverFlag = gameOverFlag
    }

    /** Resets counters for a new run while keeping the same screen instance. */
    fun resetForNewRun() {
        updateCount = 0f
        enemyCounter = INITIAL_ENEMY_COUNTER
        score = 0
        scoreUploaded = false
        speed = 0
        gameOverFlag = false
    }

    companion object {
        private const val INITIAL_ENEMY_COUNTER = 300f
    }
}
